### DISCUSSION TITLE
How do I install "playsound" in Python?

### DISCUSSION THREAD

**Discussion by frozen-cdn** [termux/termux-app#3306] [Q&A]
Hi all, I have a python script I want to use that plays a sound file at some point.  It looks like it calls for something called playsound, which fails to install:

` $ pip install playsound
Collecting playsound
  Using cached playsound-1.3.0.tar.gz (7.7 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
 
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [29 lines of output]
      Traceback (most recent call last):
        File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/tmp/pip-build-env-pa_2_lv3/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 338, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/tmp/pip-build-env-pa_2_lv3/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 320, in _get_build_requires
          self.run_setup()
        File "/data/data/com.termux/files/usr/tmp/pip-build-env-pa_2_lv3/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 485, in run_setup
          self).run_setup(setup_script=setup_script)                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/tmp/pip-build-env-pa_2_lv3/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 335, in run_setup
          exec(code, locals())
        File "<string>", line 6, in <module>
        File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1262, in getsource
          lines, lnum = getsourcelines(object)
                        ^^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1244, in getsourcelines
          lines, lnum = findsource(object)
                        ^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1081, in findsource
          raise OSError('could not get source code')
      OSError: could not get source code
      [end of output]
 
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error
 
× Getting requirements to build wheel did not run successfully.
│ exit code: 1                                          ╰─> See above for output.
 
note: This error originates from a subprocess, and is likely not a problem with pip.`

Would love assistance with this, thanks very much!

**Comment by riverallzero** [comment]
Hi @frozen905 , i hope my answer will be helpful😀

First, Upgrade your pip version to the latest version by running the command **"pip install --upgrade pip"** and retry it.
If it still doesn't work out, install the required build dependencies for playsound by running the command **"pip install build-essential libasound2-dev"** -> Then try installing playsound again by running the command **"pip install playsound"**

**Reply by frozen-cdn** [reply]
Hi there! Thank you so much for a reply

This is all that happened, it instantly failed at the first instruction unfortunately

`
~ $ python playsoundtest.py

Traceback (most recent call last):
  File "/data/data/com.termux/files/home/playsoundtest.py", line 5, in <module>
    
from playsound import playsound
ModuleNotFoundError: No module named 'playsound'

~ $ pip install --upgrade pip

ERROR: Installing pip is forbidden, this will break the python-pip package (termux).

~ $`

**Reply by riverallzero** [reply]
Oh you used Termux, sorry 😂 
Please trying this `pkg install python3-playground`

**Reply by frozen-cdn** [reply]
Hi again!

No such package is the result of that command. And just in case python3-playground was a typo and was supposed to be python3-playsound, that failed as well unfortunately.  Yep, using Termux, I meant to post this under the Termux project hopefully 😳

`E: Unable to locate package python3-playground`

`E: Unable to locate package python3-playsound`

**Reply by riverallzero** [reply]
hi, install the independece package and reinstall playsound!!
`$ pkg install libpulse libavcodec libavformat libavutil libswscale` 
`$ pip install playsound`

**Reply by frozen-cdn** [reply]
None of that worked at all unfortunately, not even one of them 

[code]
~ $ pkg install libpulse libavcodec libavformat libavutil libswscale
No mirror or mirror group selected. You might want to select one by running 'termux-change-repo'
Checking availability of current mirror:
[*] https://packages-cf.termux.dev/apt/termux-main: ok
Get:1 https://packages-cf.termux.dev/apt/termux-main stable InRelease [14.0 kB]
Hit:2 https://packages-cf.termux.dev/apt/termux-x11 x11 InRelease
Get:3 https://packages-cf.termux.dev/apt/termux-main stable/main aarch64 Packages [496 kB]
Fetched 510 kB in 2s (249 kB/s)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
39 packages can be upgraded. Run 'apt list --upgradable' to see them.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package libpulse
E: Unable to locate package libavcodec
E: Unable to locate package libavformat
E: Unable to locate package libavutil
E: Unable to locate package libswscale
~ $
[/code]

**Reply by riverallzero** [reply]
hard issue to solve🥹
this is my last solution i can handle... 
- fist, install package 
`pkg install python`
`pkg install clang`
`pkg install ffmpeg libavcodec-extra`
- then reinstall playsound
`pip install playsound`

**Reply by EntropicBlackhole** [reply]
Hey @frozen905 

Did you manage to get it working? If so I would be extremely thankful

**Reply by frozen-cdn** [reply]
> Hey @frozen905
> 
> Did you manage to get it working? If so I would be extremely thankful

Unfortunately not. Termux has no support that I've been able to find. Nothing mentioned above works, at all

If you ever fix it I'd love to know! 🙏🏻

**Reply by EntropicBlackhole** [reply]
I actually just did! What I did was install it from source; search up the repository for playsound, git clone it somewhere, then go the directory and run `python setup.py install`

This worked for me, building it from source, it was surprisingly fast

I'm using it for running AutoGPT on my phone, which is useful since I'd use it on the go!

**Reply by frozen-cdn** [reply]
Thanks for the instructions! I completed this okay as you said.  Then installed pygobjects as it requests. However, after trying the actual playsound function, I get a page full of errors

```py
>>> from playsound import playsound
>>> playsound('test.wav')
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py", line 244, in <module>
    playsound(argv[1])
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py", line 146, in _playsoundNix
    gi.require_version('Gst', '1.0')
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/gi/__init__.py", line 126, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace Gst not available
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py", line 237, in <lambda>
    playsound = lambda sound, block = True: _playsoundAnotherPython('/usr/bin/python3', sound, block, macOS = False)
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py", line 212, in _playsoundAnotherPython
    t.join()
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py", line 201, in join
    raise self.exc
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py", line 194, in run
    self.ret = self._target(*self._args, **self._kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py", line 209, in <lambda>
    t = PropogatingThread(target = lambda: check_call([otherPython, playsoundPath, _handlePathOSX(sound) if macOS else sound]))
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', '/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py', 'test.wav']' returned non-zero exit status 1.
>>>
```

Does yours actually work when you try it?

Thanks again!

**Reply by frozen-cdn** [reply]
Just to follow up. Python Discord informed me it was related to something called gstreamer, which I was able to 'pkg install gstreamer' with no problems.

Error output is drastically reduced to just 1 error now:
```py
>>> from playsound import playsound
>>> playsound('test.wav')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/playsound-1.3.0-py3.11.egg/playsound.py", line 158, in _playsoundNix
    playbin.props.uri = 'file://' + pathname2url(path)
    ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'props'
>>>
```

**Reply by shreyaspawar05** [reply]
Hey there, 
I had this problem as well and came across a solution as,
Firstly you need to upgrade setuptools wheel 

step 1. command - ( pip install --upgrade setuptools wheel )
step 2. command - ( pip install playsound )

Hopefully, this will def fix it!!

**Reply by morningu** [reply]
@shreyaspawar05 
thanks. your recommendation 'step 1.' helped me.

Before I command 'step 1.', I typed in following command.
- ( python.exe -m pip install --upgrade pip )

thanks again. :)

**Reply by abhinavganji** [reply]
Thank you so much

**Reply by kusrohit** [reply]
@shreyaspawar05 
Thanks buddy !
Step 1 works well for me😊

**Reply by Shrawan662000** [reply]
@shreyaspawar05 Thanks a lot brother, its worked for me

**Reply by elizeubarbosaabreu** [reply]
> pip install --upgrade setuptools wheel

Obrigado. Salvou meu dia

**Comment by marwandhiaurrahman** [comment]
i just run it , and solved

`pip install --upgrade wheel`

**Reply by LorenzoProjex** [reply]
So simple, yet so effective. This worked, thanks for the help!

**Reply by Leomconti** [reply]
Great solution. Simple. Worked

**Reply by frozen-cdn** [reply]
This didn't work for me, no

```py
$ pip install playsound
Collecting playsound
  Downloading playsound-1.3.0.tar.gz (7.7 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [29 lines of output]
      Traceback (most recent call last):
        File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/tmp/pip-build-env-unqonztn/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 341, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/tmp/pip-build-env-unqonztn/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 323, in _get_build_requires
          self.run_setup()
        File "/data/data/com.termux/files/usr/tmp/pip-build-env-unqonztn/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 488, in run_setup
          self).run_setup(setup_script=setup_script)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/tmp/pip-build-env-unqonztn/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 338, in run_setup
          exec(code, locals())
        File "<string>", line 6, in <module>
        File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1262, in getsource
          lines, lnum = getsourcelines(object)
                        ^^^^^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1244, in getsourcelines
          lines, lnum = findsource(object)
                        ^^^^^^^^^^^^^^^^^^
        File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1081, in findsource
          raise OSError('could not get source code')
      OSError: could not get source code
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
~ $ pip install --upgrade wheel
Requirement already satisfied: wheel in /data/data/com.termux/files/usr/lib/python3.11/site-packages (0.40.0)
```

**Reply by breadgirlxoxo** [reply]
best solution out here

**Comment by Bugsoul** [comment]
Firstly, you have to install this;

pip install setuptools

**Reply by kodzonko** [reply]
yup, that worked for me (python 3.12.6, pip up-tp-date, wheel updated - didn't help, on PopOS)

**Comment by jbworkdevelep** [comment]
pip install --upgradewheel

pip install playsound 

worked for me!🙏👍👍

![Screenshot_2023-07-04_11_51_17](https://github.com/termux/termux-app/assets/32499879/ae7eb43d-ff5c-4424-8e6e-1393e3e19da4)
![Screenshot_2023-07-04_11_51_33](https://github.com/termux/termux-app/assets/32499879/2e448310-b73c-4736-9f68-1947486cca7b)

**Comment by shreyaspawar05** [comment]
> Hi all, I have a python script I want to use that plays a sound file at some point. It looks like it calls for something called playsound, which fails to install:
> 
> ` $ pip install playsound Collecting playsound Using cached playsound-1.3.0.tar.gz (7.7 kB) Installing build dependencies ... done Getting requirements to build wheel ... error error: subprocess-exited-with-error
> 
> × Getting requirements to build wheel did not run successfully. │ exit code: 1 ╰─> [29 lines of output] Traceback (most recent call last): File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in main() File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main json_out['return_val'] = hook(**hook_input['kwargs']) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel return hook(config_settings) ^^^^^^^^^^^^^^^^^^^^^ File "/data/data/com.termux/files/usr/tmp/pip-build-env-pa_2_lv3/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 338, in get_requires_for_build_wheel return self._get_build_requires(config_settings, requirements=['wheel']) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/data/data/com.termux/files/usr/tmp/pip-build-env-pa_2_lv3/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 320, in _get_build_requires self.run_setup() File "/data/data/com.termux/files/usr/tmp/pip-build-env-pa_2_lv3/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 485, in run_setup self).run_setup(setup_script=setup_script) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/data/data/com.termux/files/usr/tmp/pip-build-env-pa_2_lv3/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 335, in run_setup exec(code, locals()) File "", line 6, in File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1262, in getsource lines, lnum = getsourcelines(object) ^^^^^^^^^^^^^^^^^^^^^^ File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1244, in getsourcelines lines, lnum = findsource(object) ^^^^^^^^^^^^^^^^^^ File "/data/data/com.termux/files/usr/lib/python3.11/inspect.py", line 1081, in findsource raise OSError('could not get source code') OSError: could not get source code [end of output]
> 
> note: This error originates from a subprocess, and is likely not a problem with pip. error: subprocess-exited-with-error
> 
> × Getting requirements to build wheel did not run successfully. │ exit code: 1 ╰─> See above for output.
> 
> note: This error originates from a subprocess, and is likely not a problem with pip.`
> 
> Would love assistance with this, thanks very much!


step 1. command - ( pip install --upgrade setuptools wheel )
step 2. command - ( pip install playsound )

Hopefully, this will def fix it!!

**Reply by Kunalp9** [reply]
Thanks mate. It worked for me!

**Reply by Devash-beatox** [reply]
this helped me

**Comment by Roscpy** [comment]
Comment installer playsound dans termux
